### PR TITLE
Fix favourites bug #151 and #165

### PR DESCRIPTION
**Pull request by simhayoz**

### CODE REVIEW

**Comment by simhayoz** [comment]
**For bug #151:**
It was due to the `MultipleCallHandler` that didn't increment the counter of the number of data it receives if a `onFailure` was called. Thus never reaching the max number of data to retrieve.
It is now changed to be more parameterized, so we can do something on failure in a `MultipleCallHandler` with the `Consumer<MultipleCallHandler> doOnSingleFailure` and we can also choose whether the counter should be incremented or not with a variable.

**Comment by simhayoz** [comment]
**For bug #165:**
It was probably due to the `onBindViewHolder` from `RecipeMiniatureAdapter`, as the method `onBindViewHolder` is called everytime it needs to draw another recipe. Thus the second time it is called for any recipe, we call `setChecked` from `ToggleButton` for a button that already has a `OnCheckedChangeListener`, thus it would call `setOrRemoveRecipe` from `FavouritesUtils` that would update the user in database (this alone would just cause a lot of updates on Firebase with the same data).
But the problem is, when another user post a recipe, the next time that `RecipeMiniatureAdapter` will update for the user on HomePage (when the user scrolls down), then every recipe will go down by one in the `RecyclerView` to let the space for the newly created recipe on top. 
Thus every `ToggleButton` but the last will be associated with a recipe that is not the right one and they will all call `setChecked` (as explained above). This will change the favourites from what was found in another recipe, thus making the favourites disappear and sometimes change for another recipe in database.

**TL;DR:**
A problem in `onBindViewHolder` was forcing an update to the user everytime a user scroll on HomePage. A new recipe in database was causing an offset between recipes and favourite buttons and with the update, was deleting the favourites.

**Review by wilderino — APPROVED** [review]

**Review by ChapatteMateo — APPROVED** [review]
